mmdet/models/necks/bifpn.py

```diff
@@ -0,0 +1,323 @@
+
+        for node in self.nodes:
+            output.append(node(output))
+
+        return tuple(output[-self.num_outs::])
+
+
+@NECKS.register_module
+class BiFPN(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 input_indices,
+                 num_layers,
+                 num_outs,
+                 strides,
+                 weight_method="fast_attn",
+                 act_cfg="silu",
+                 separable_conv=True,
+                 epsilon=0.0001,
+                 reduction_ratio = 2.0):
+        super(BiFPN, self).__init__()
+        assert isinstance(in_channels, list)
+        self.num_backbone_features = len(in_channels)
+
+        assert self.num_backbone_features >= 2
+        assert num_outs - self.num_backbone_features >= 2
+        assert len(input_indices) == len(strides)
+
+        self.in_channels = in_channels
+        self.num_outs = num_outs
+
+        self.extra_convs = nn.ModuleList()
+
+        min_level = input_indices[0]
+        max_level = num_outs + min_level - 1
+
+        # input_offsets are the nodes that the current node is getting its inputs from (either 2 or 3).
+        input_offsets = []
+        # feature_reduction is the reduction values from the inputted nodes coming from the backbone
+        feature_reduction = strides
+        # reduction is the reduction values for the neck nodes
+        reduction = []
+
+        node_ids = {min_level + i: [i] for i in range(num_outs)}
+
+        level_last_id = lambda level: node_ids[level][-1]
+        level_all_ids = lambda level: node_ids[level]
+        id_cnt = itertools.count(num_outs)
+
+        for i in range(max_level - 1, min_level - 1, -1):
+            # top-down path
+            reduction.append(1 << i)
+            input_offsets.append([level_last_id(i), level_last_id(i + 1)])
+            node_ids[i].append(next(id_cnt))
+
+        for i in range(min_level + 1, max_level + 1):
+            # bottom-up path
+            reduction.append(1 << i)
+            input_offsets.append(level_all_ids(i) + [level_last_id(i - 1)])
+            node_ids[i].append(next(id_cnt))
+
+        for i in range(self.num_outs - self.num_backbone_features):
+            if i == 0:
+                input_channels = in_channels[-1]
+            else:
+                input_channels = out_channels
+            if input_channels != out_channels:
+                self.extra_convs.append(
+                    nn.Sequential(
+                        ConvModule(input_channels, out_channels, kernel_size=1,
+                                   norm_cfg=dict(type='BN'), act_cfg=None),
+                        nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
+                    )
+                )
+            else:
+                self.extra_convs.append(nn.MaxPool2d(kernel_size=3, stride=2, padding=1))
+            feature_reduction.append(int(feature_reduction[-1] * reduction_ratio))
```

Comment:
should not `reduction_ratio` be int always?

Reply:
Similar code is done from the [code](https://github.com/rwightman/efficientdet-pytorch/blob/645d84a6f0cd837703f98f48179a06c354902515/effdet/efficientdet.py#L332) I based much of this work on. In this implementation, `reduction_ratio` is a calculated value that could have a decimal